File: Day4/solution.ex
defmodule Solution do
  def to_ranges(list) do
    list
    |> Enum.map(fn x -> String.split(x, "-") |> Enum.map(&String.to_integer/1) end)
    |> Enum.map(fn [s, e] -> Enum.to_list(s..e) |> MapSet.new() end)
  end

  def input(filename) do
    File.read!(filename)
    |> String.split("\n", trim: true)
    |> Enum.map(&String.split(&1, ","))
    |> Enum.map(&to_ranges/1)
  end

  def part_one(data) do
    data
    |> Enum.map(fn [x, y] -> MapSet.intersection(x, y) |> then(&Enum.member?([x, y], &1)) end)
    |> Enum.count(fn over -> over == true end)
  end

  def part_two(data) do
    data
    |> Enum.map(fn [x, y] -> MapSet.intersection(x, y) end)
    |> Enum.count(fn range -> not Enum.empty?(range) end)
  end

  def main do
    data = input("input.txt")
    IO.puts "Part one: #{part_one(data)}"
    IO.puts "Part two: #{part_two(data)}"
  end
end

File: Day4/solution.hs
import Data.List.Split (splitOn)
import Data.List (intersect)

toRange :: String -> [Int]
toRange s = [x..y]
  where [x, y] = map read $ splitOn "-" s

input :: FilePath -> IO [[[Int]]]
input f = map (map toRange . splitOn ",") . filter (/= "") . splitOn "\n" <$> readFile f

partOne :: [[[Int]]] -> Int
partOne = length . filter (== True) . map overlaps
  where
    overlaps [x, y] = let i = intersect x y in i == x || i == y
    overlaps _ = error "Invalid input"

partTwo :: [[[Int]]] -> Int
partTwo = length . filter (== True) . map overlaps
  where
    overlaps [x, y] = not $ null (x `intersect` y)
    overlaps _ = error "Invalid input"

main :: IO ()
main = do
  d <- input "input.txt"
  putStrLn $ "Part one: " ++ show (partOne d)
  putStrLn $ "Part two: " ++ show (partTwo d)
